rand_os/src/solarish.rs

```diff
@@ -110,18 +112,50 @@ impl OsRngImpl for OsRng {
     }
 }
 
-fn getrandom(buf: &mut [u8], blocking: bool) -> libc::c_long {
-    extern "C" {
-        fn syscall(number: libc::c_long, ...) -> libc::c_long;
-    }
+#[cfg(target_os = "illumos")]
+type GetRandomFn = unsafe extern fn(*mut u8, libc::size_t, libc::c_uint)
+    -> libc::ssize_t;
+#[cfg(target_os = "solaris")]
+type GetRandomFn = unsafe extern fn(*mut u8, libc::size_t, libc::c_uint)
+    -> libc::c_int;
+
+// Use dlsym to determine if getrandom(2) is present in libc.  On Solarish
+// systems, the syscall interface is not stable and can change between
+// updates.  Even worse, issuing unsupported syscalls will cause the system
+// to generate a SIGSYS signal (usually terminating the program).
+// Instead the stable APIs are exposed via libc.  Cache the result of the
+// lookup for future calls.  This is loosely modeled after the
+// libstd::sys::unix::weak macro which unfortunately is not exported.
+fn fetch() -> Option<&'static GetRandomFn> {
+    static FPTR: AtomicUsize = AtomicUsize::new(1);
 
-    const SYS_GETRANDOM: libc::c_long = 143;
+    unsafe {
+       if FPTR.load(Ordering::SeqCst) == 1 {
+            let name = "getrandom\0";
+            let addr = libc::dlsym(libc::RTLD_DEFAULT,
+                                   name.as_ptr() as *const _) as usize;
+            FPTR.store(addr, Ordering::SeqCst);
+       }
+
+       if FPTR.load(Ordering::SeqCst) == 0 {
+           return None;
+       } else {
+           mem::transmute::<&AtomicUsize, Option<&GetRandomFn>>(&FPTR)
+       }
+   }
+}
+
+fn getrandom(buf: &mut [u8], blocking: bool) -> libc::ssize_t {
     const GRND_NONBLOCK: libc::c_uint = 0x0001;
     const GRND_RANDOM: libc::c_uint = 0x0002;
 
-    unsafe {
-        syscall(SYS_GETRANDOM, buf.as_mut_ptr(), buf.len(),
-                if blocking { 0 } else { GRND_NONBLOCK } | GRND_RANDOM)
+    if let Some(rand) = fetch() {
+        unsafe {
+            rand(buf.as_mut_ptr(), buf.len(),
+                 if blocking { 0 } else { GRND_NONBLOCK } | GRND_RANDOM) as libc::ssize_t
```

Comment:
Would be nicer to move the `if` block outside the `unsafe` via a `let` binding.

Reply:
I've cleaned it up, though I've not been able to get any combination of pointer casting to work -- the given suggestion segfaults.  Usually it ends up that it won't let the cast from usize->GetRandomFn in whatever form (and suggests using ::From).

All of the examples I've been able to find using libc::dlsym end up using mem::transmute, so I'm not sure if there is currently anything better.